n64/src/ap.c

```diff
@@ -1141,13 +1141,75 @@ void ap_update() {
 
 void ap_check_enough_notes(u16 start, u16 end) {
   if (start == end || start > end) return;
+
+  const char* names[] = {
+    "EGG AIM",
+    "BREEGULL BLASTER",
+    "GRIP GRAB",
+    "FIRE EGGS",
+    "BILL DRILL",
+    "BEAK BAYONET",
+    "GRENADE EGGS",
+    "SPLIT UP",
+    "PACK WHACK",
+    "AIRBORNE EGG AIMING",
+    "ICE EGGS",
+    "WING WHACK",
+    "SUB-AQUA EGG AIMING",
+    "TALON TORPEDO",
+    "CLOCKWORK KAZOOIE EGGS",
+    "SPRINGY STEP SHOES",
+    "TAXI PACK",
+    "HATCH",
+    "CLAW CLAMBER BOOTS",
+    "SNOOZE PACK",
+    "LEG SPRING",
+    "SHACK PACK",
+    "GLIDE",
+    "SACK PACK"
+  };
+
+  const int UNLOCK_LIMIT = 5;
+  int unlocks[UNLOCK_LIMIT];
+  int n_unlocks = 0;
+
   for (int i = 0; i < sizeof(ap_memory.pc.settings.silo_requirements)/sizeof(*ap_memory.pc.settings.silo_requirements); i++) {
     u16 amount = ap_memory.pc.settings.silo_requirements[i];
     if (start < amount && end >= amount) {
-      ap.internal_icon = BT_ZOOMBOX_ICON_JAMJARS;
-      strcpy(ap.internal_message, "YOU HAVE ENOUGH NOTES FOR A NEW MOVE!");
-      break;
+      if (n_unlocks < UNLOCK_LIMIT) {
+        unlocks[n_unlocks] = i;
+      }
+      ++n_unlocks;
+    }
+  }
+
+  if (n_unlocks == 0) {
+    return;
+  }
+
+  ap.internal_icon = BT_ZOOMBOX_ICON_JAMJARS;
+
+  if (n_unlocks > UNLOCK_LIMIT) {
+    // Too many silos unlocked at the same time.
+    strcpy(ap.internal_message, "YOU HAVE ENOUGH NOTES TO CHECK MULTIPLE SILOS!");
+    return;
+  }
+
+  strcpy(ap.internal_message, "YOU HAVE ENOUGH NOTES TO CHECK THE ");
+
+  if (n_unlocks == 1) {
+    strcat(ap.internal_message, names[unlocks[0]]);
+    strcat(ap.internal_message, " SILO!");
+  }
+  else {
+    for (int i = 0; i < n_unlocks - 1; ++i){
+      strcat(ap.internal_message, names[unlocks[i]]);
+      strcat(ap.internal_message, ", ");
     }
+
+    strcat(ap.internal_message, "AND ");
+    strcat(ap.internal_message, names[unlocks[n_unlocks - 1]]);
+    strcat(ap.internal_message, " SILOS!");
   }
```

Comment:
```suggestion
  for (int i = 0; i < n_unlocks; i++) {
    strcat(ap.internal_message, names[unlocks[i]]);
    if (n_unlocks == 1) strcat(ap.internal_message, " SILO!"); // Only one
    else if (i+1 == n_unlocks) strcat(ap.internal_message, " SILOS!"); // Last one
    else if (i+2 == n_unlocks) strcat(ap.internal_message, " AND "); // Second to last
    else strcat(ap.internal_message, ", ");
  }
```
It looks like your code would generate sentences like: "...CHECK THE EGG AIM, AND GRIP GRAB SILOS!"
I think having ", AND" is wrong when there are only 2 items, but is otherwise optional.
This code should always use " AND " instead and should be functionally the same otherwise.
It is restructured to be more readable (in my opinion), and to not have 3 different areas where the name is appended.

It's also worth mentioning that `strcat` becomes slower the longer a string is. It really makes no difference here in this context since this is run rarely.

Reply:
I'm aware of strcat being slower, I don't think this is close to being a bottleneck here. Happy with the rest though.